File: lib/vbrain/git.rb
require "open3"
require "fileutils"
require_relative "paths"

module VBrain
  module Git
    # O SQLite (db/vbrain.sqlite3) é um índice descartável — pode ser apagado e
    # reconstruído com scripts/reindex.rb — mas é versionado por conveniência
    # (clone/pull já trazem o índice pronto, sem precisar reindexar). Por isso
    # /db/ NÃO entra no .gitignore. Só staging volátil e lixo de SO ficam fora.
    GITIGNORE = <<~IGN.freeze
      /raw/.tmp/
      .DS_Store
    IGN

    class Error < StandardError; end

    def self.repo_initialized?(dir = Paths.data_home)
      File.directory?(File.join(dir, ".git"))
    end

    def self.init!(dir = Paths.data_home)
      FileUtils.mkdir_p(dir)
      raise Error, "repo already initialized at #{dir}" if repo_initialized?(dir)

      run!(dir, "git", "init", "-b", "main")
      write_gitignore!(dir)
      run!(dir, "git", "add", ".gitignore")
      run!(dir, "git", "commit", "-m", "chore: initialize vbrain")
      dir
    end

    def self.write_gitignore!(dir = Paths.data_home)
      path = File.join(dir, ".gitignore")
      existing = File.exist?(path) ? File.read(path) : ""
      return path if existing.include?("/raw/.tmp/")

      File.write(path, GITIGNORE)
      path
    end

    def self.add_remote!(url, dir = Paths.data_home, name: "origin")
      run!(dir, "git", "remote", "add", name, url)
    end

    def self.has_remote?(name: "origin", dir: Paths.data_home)
      out, _err, status = Open3.capture3("git", "remote", "get-url", name, chdir: dir)
      status.success? && !out.strip.empty?
    end

    def self.current_branch(dir = Paths.data_home)
      out, _err, status = Open3.capture3("git", "rev-parse", "--abbrev-ref", "HEAD", chdir: dir)
      status.success? ? out.strip : nil
    end

    def self.changes?(dir = Paths.data_home)
      out, _err, status = Open3.capture3("git", "status", "--porcelain", chdir: dir)
      status.success? && !out.strip.empty?
    end

    def self.commit!(message, dir = Paths.data_home)
      run!(dir, "git", "add", "-A")
      return { "committed" => false, "reason" => "no changes" } unless changes_staged?(dir)

      run!(dir, "git", "commit", "-m", message)
      sha = `git -C #{dir.shellescape} rev-parse HEAD`.strip
      { "committed" => true, "sha" => sha, "message" => message }
    end

    def self.changes_staged?(dir = Paths.data_home)
      _out, _err, status = Open3.capture3("git", "diff", "--cached", "--quiet", chdir: dir)
      !status.success?
    end

    def self.push!(dir = Paths.data_home, name: "origin", branch: nil)
      return { "pushed" => false, "reason" => "no remote" } unless has_remote?(name: name, dir: dir)

      branch ||= current_branch(dir)
      run!(dir, "git", "push", "-u", name, branch)
      { "pushed" => true, "remote" => name, "branch" => branch }
    end

    def self.run!(dir, *cmd)
      out, err, status = Open3.capture3(*cmd, chdir: dir)
      raise Error, "#{cmd.join(' ')} failed: #{err.strip}\n#{out.strip}" unless status.success?

      out
    end
  end
end

require "shellwords"

File: internal/git/git.go
// Package git embrulha as operações git do vbrain (init, commit, push) com dois
// backends selecionados em runtime: o git do sistema quando presente (usa as
// credenciais já configuradas do usuário no push), senão go-git puro-Go (sem
// dependência externa; push usa um PAT coletado no install).
//
// Operações de leitura (branch, remote, changes) usam um único caminho go-git —
// elas leem um repositório git padrão independentemente de quem o criou. Porta
// de lib/vbrain/git.rb. O índice SQLite é versionado de propósito: /db/ NÃO
// entra no .gitignore.
package git

import (
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	gogit "github.com/go-git/go-git/v5"
)

// Gitignore: só staging volátil e lixo de SO. /db/ fica versionado.
const Gitignore = "/raw/.tmp/\n.DS_Store\n"

// CommitResult espelha o hash retornado por Git.commit! no Ruby.
type CommitResult struct {
	Committed bool   `json:"committed"`
	Reason    string `json:"reason,omitempty"`
	SHA       string `json:"sha,omitempty"`
	Message   string `json:"message,omitempty"`
}

// PushResult espelha o hash retornado por Git.push!.
type PushResult struct {
	Pushed bool   `json:"pushed"`
	Reason string `json:"reason,omitempty"`
	Remote string `json:"remote,omitempty"`
	Branch string `json:"branch,omitempty"`
}

// backend abstrai as operações que mutam o repo e dependem de credenciais.
type backend interface {
	Init(dir string) error
	Commit(message, dir string) (CommitResult, error)
	Push(dir, name, branch string) (PushResult, error)
	AddRemote(url, dir, name string) error
}

// systemGitAvailable detecta o git no PATH; é uma var para os testes poderem
// forçar o backend go-git.
var systemGitAvailable = func() bool {
	_, err := exec.LookPath("git")
	return err == nil
}

func selected() backend {
	if systemGitAvailable() {
		return systemBackend{}
	}
	return gogitBackend{}
}

// BackendName devolve "system" ou "gogit" — útil pro install reportar e decidir
// se precisa coletar PAT.
func BackendName() string {
	if systemGitAvailable() {
		return "system"
	}
	return "gogit"
}

// Init cria o repo (branch main), escreve o .gitignore e faz o commit inicial.
func Init(dir string) error { return selected().Init(dir) }

// Commit faz o stage de tudo e commita; no-op se nada mudou.
func Commit(message, dir string) (CommitResult, error) { return selected().Commit(message, dir) }

// Push faz push da branch; no-op se não houver remote.
func Push(dir, name, branch string) (PushResult, error) { return selected().Push(dir, name, branch) }

// AddRemote adiciona um remote nomeado.
func AddRemote(url, dir, name string) error { return selected().AddRemote(url, dir, name) }

// --- operações de leitura (backend-agnósticas, via go-git) ---

// RepoInitialized indica se dir já tem um repositório git.
func RepoInitialized(dir string) bool {
	fi, err := os.Stat(filepath.Join(dir, ".git"))
	return err == nil && fi.IsDir()
}

// WriteGitignore escreve o .gitignore se ainda não contiver a marca; idempotente.
func WriteGitignore(dir string) (string, error) {
	path := filepath.Join(dir, ".gitignore")
	if b, err := os.ReadFile(path); err == nil && strings.Contains(string(b), "/raw/.tmp/") {
		return path, nil
	}
	return path, os.WriteFile(path, []byte(Gitignore), 0o644)
}

// CurrentBranch devolve a branch atual, ou "" se indeterminada.
func CurrentBranch(dir string) string {
	repo, err := gogit.PlainOpen(dir)
	if err != nil {
		return ""
	}
	head, err := repo.Head()
	if err != nil {
		return ""
	}
	return head.Name().Short()
}

// HasRemote indica se o remote nomeado existe com URL.
func HasRemote(dir, name string) bool {
	repo, err := gogit.PlainOpen(dir)
	if err != nil {
		return false
	}
	r, err := repo.Remote(name)
	return err == nil && len(r.Config().URLs) > 0
}

// Changes indica se há mudanças no working tree.
func Changes(dir string) bool {
	repo, err := gogit.PlainOpen(dir)
	if err != nil {
		return false
	}
	wt, err := repo.Worktree()
	if err != nil {
		return false
	}
	st, err := wt.Status()
	if err != nil {
		return false
	}
	return !st.IsClean()
}
